In [1]:
import numpy as np
import pandas as pd

from warnings import filterwarnings
filterwarnings("ignore")

In [2]:
amazon = pd.read_csv("../data/amz_com_ecommerce_sample.csv", encoding = "ISO-8859-1")
flipkart = pd.read_csv("../data/flipkart_com_ecommerce_sample.csv", encoding = "ISO-8859-1")

In [3]:
amazon.sample(5)

uniq_id            crawl_timestamp  \
1050   bf772feb47359879cc8adbe65610215e  2016-03-03 23:09:59 +0000   
14674  d09c79ea4b2dd3659d7ede8161ff4ea1  2015-12-12 11:46:53 +0000   
1339   f170abb4e3c86d8f0e6c3b4ecbec9eb1  2016-03-25 15:39:25 +0000   
19904  575b6a101ccb86e5df3bb42f4bd2ad63  2015-12-01 10:15:43 +0000   
7417   24dd01395dcf8b4bb9570906b9877c65  2015-12-01 06:13:00 +0000   

                                             product_url  \
1050   http://www.flipkart.com/wolfer-feather-weight-...   
14674  http://www.flipkart.com/stanley-93-021-22-pipe...   
1339   http://www.flipkart.com/fourgee-slim-fit-girl-...   
19904  http://www.flipkart.com/walldesign-small-vinyl...   
7417   http://www.flipkart.com/dressberry-crystal-ena...   

                                            product_name  \
1050   Wolfer Feather Weight (Right Hand) Batting Glo...   
14674                      Stanley 93-021-22 Pipe Cutter   
1339                       fourgee Slim Fit Girl's Jeans   
19904                         WALLDESIGN S VINYL STICKER   
7417     DressBerry Crystal Enamel Plated Metal Necklace   

                                   product_category_tree               pid  \
1050   ["Sports & Fitness >> Team Sports >> Cricket >...  SGEEG57EZYEGGNMN   
14674  ["Tools & Hardware >> Tools >> Hand Tools >> C...  CUTDTXFGNMHJRUHX   
1339   ["Clothing >> Kids' Clothing >> Girls Wear >> ...  JEAEGY7XWTDDWDEQ   
19904  ["Baby Care >> Baby & Kids Gifts >> Stickers >...  STIE7KFGSBYBQG8U   
7417    ["Jewellery >> Necklaces & Chains >> Necklaces"]  NKCEBHGBYC9ZH3ZX   

       retail_price  discounted_price  \
1050           2394              1998   
14674          1197              1375   
1339           1995               667   
19904          1064               600   
7417            698               356   

                                                   image  \
1050   ["http://img5a.flixcart.com/image/sport-glove/...   
14674  ["http://img6a.flixcart.com/image/cutter/u/h/x...   
1339   ["http://img6a.flixcart.com/image/jean/5/y/7/b...   
19904  ["http://img6a.flixcart.com/image/wall-decorat...   
7417   ["http://img5a.flixcart.com/image/necklace-cha...   

       is_FK_Advantage_product  \
1050                     False   
14674                    False   
1339                     False   
19904                    False   
7417                     False   

                                             description       product_rating  \
1050   Specifications of Wolfer Feather Weight (Right...  No rating available   
14674  Buy Stanley 93-021-22 Pipe Cutter for Rs.1087 ...  No rating available   
1339   Key Features of fourgee Slim Fit Girl's Jeans ...  No rating available   
19904  Buy WallDesign Small Vinyl Sticker for Rs.530 ...  No rating available   
7417   DressBerry Crystal Enamel Plated Metal Necklac...  No rating available   

            overall_rating       brand  \
1050   No rating available      Wolfer   
14674  No rating available     Stanley   
1339   No rating available     fourgee   
19904  No rating available  WallDesign   
7417   No rating available  DressBerry   

                                  product_specifications  
1050   {"product_specification"=>[{"key"=>"Sport Type...  
14674  {"product_specification"=>[{"key"=>"Body Mater...  
1339   {"product_specification"=>[{"key"=>"Closure", ...  
19904  {"product_specification"=>[{"key"=>"Number of ...  
7417   {"product_specification"=>[{"key"=>"Collection...

In [4]:
flipkart.sample(5)

uniq_id            crawl_timestamp  \
11200  f9a8320c182841024f47ebcba284085a  2015-12-31 09:19:31 +0000   
15358  269ab7a6d1406297d073221da4b86465  2015-12-12 11:46:53 +0000   
19140  53155a174cf6ba92df741c1b00e931df  2016-03-02 19:09:22 +0000   
12694  db548ae102f736b498e1d59c5f7be8a7  2015-12-30 00:17:46 +0000   
6856   4c449766e715bc41023244c9042e0a82  2015-12-01 06:13:00 +0000   

                                             product_url  \
11200  http://www.flipkart.com/portia-polka-print-tre...   
15358  http://www.flipkart.com/suvarnadeep-rose-rhodi...   
19140  http://www.flipkart.com/czds-india-shoulder-ba...   
12694  http://www.flipkart.com/ausy-women-s-solid-cas...   
6856   http://www.flipkart.com/vincent-valentine-pari...   

                                            product_name  \
11200                      Portia Polka Print Trendy Cap   
15358  Suvarnadeep Rose Rhodium Zircon Sterling Silve...   
19140                            Czds India Shoulder Bag   
12694                    Ausy Women's Solid Casual Shirt   
6856   Vincent Valentine Paris Set of Dark Fire & Dar...   

                                   product_category_tree               pid  \
11200  ["Baby Care >> Infant Wear >> Baby Girls' Clot...  CAPEDKXZJYCYQANZ   
15358                ["Jewellery >> Pendants & Lockets"]  PELE58U3FFP8EAAT   
19140  ["Bags, Wallets & Belts >> Bags >> Hand Bags >...  HMBEGFGAH9ZS8HZH   
12694  ["Clothing >> Women's Clothing >> Western Wear...  SHTEC8AHDXHVAEQW   
6856   ["Beauty and Personal Care >> Fragrances >> De...  CAGEYUXVF2FNARKD   

       retail_price  discounted_price  \
11200         900.0             399.0   
15358        1550.0            1550.0   
19140         599.0             299.0   
12694         999.0             487.0   
6856          355.0             310.0   

                                                   image  \
11200  ["http://img6a.flixcart.com/image/cap/a/n/z/pi...   
15358  ["http://img6a.flixcart.com/image/pendant-lock...   
19140  ["http://img6a.flixcart.com/image/hand-messeng...   
12694  ["http://img6a.flixcart.com/image/shirt/e/q/w/...   
6856   ["http://img5a.flixcart.com/image/combo-gift-s...   

       is_FK_Advantage_product  \
11200                    False   
15358                     True   
19140                    False   
12694                    False   
6856                     False   

                                             description       product_rating  \
11200  Portia Polka Print Trendy Cap - Buy Multicolor...  No rating available   
15358  Suvarnadeep Rose Rhodium Zircon Sterling Silve...                    3   
19140  Key Features of Czds India Shoulder Bag Ideal ...  No rating available   
12694  Ausy Women's Solid Casual Shirt - Buy Pink Aus...  No rating available   
6856   Flipkart.com: Buy Vincent Valentine Paris Set ...  No rating available   

            overall_rating        brand  \
11200  No rating available          NaN   
15358                    3  Suvarnadeep   
19140  No rating available   Czds India   
12694  No rating available         Slim   
6856   No rating available          NaN   

                                  product_specifications  
11200  {"product_specification"=>[{"key"=>"Pattern", ...  
15358  {"product_specification"=>[{"key"=>"Body Mater...  
19140  {"product_specification"=>[{"key"=>"Closure", ...  
12694  {"product_specification"=>[{"key"=>"Pattern", ...  
6856   {"product_specification"=>[{"key"=>"Number of ...

In [5]:
print(f"amazon shape: {amazon.shape} | flipkart shape: {flipkart.shape}")

amazon shape: (20000, 15) | flipkart shape: (20000, 15)


In [6]:
print(f"---- amazon null values ----\n"
      f"{amazon.isna().sum()}\n",
      
      f"\n---- flipkart null values ----\n"
      f"{flipkart.isna().sum()}"
      )

---- amazon null values ----
uniq_id                       0
crawl_timestamp               0
product_url                   0
product_name                  0
product_category_tree         0
pid                           0
retail_price                  0
discounted_price              0
image                         3
is_FK_Advantage_product       0
description                   2
product_rating                0
overall_rating                0
brand                      5864
product_specifications       14
dtype: int64
 
---- flipkart null values ----
uniq_id                       0
crawl_timestamp               0
product_url                   0
product_name                  0
product_category_tree         0
pid                           0
retail_price                 78
discounted_price             78
image                         3
is_FK_Advantage_product       0
description                   2
product_rating                0
overall_rating                0
brand                      58

## Preprocessing Datasets:

- Here I'm removing some columns that are not going to help in retrieving information.
- Rather than cleaning tons of rows; only going to select columns which are important, as mentioned in the pdf.

- Further I'm converting uppercase/capitalized words/strings into lowercase format.

In [7]:
def delete_redundant_columns(column_names: list, data: pd.DataFrame) -> None:
    
    """ removes redundant columns and returns the dataset"""
    
    if len(column_names) == 0:
        print(f"enter column names in the list to remove from the dataset.")
        return
    data.drop(columns=column_names, inplace=True)
    
    return

In [8]:
columns = ["uniq_id", "crawl_timestamp", "product_url", "product_category_tree","pid", "image", 
           "is_FK_Advantage_product", "description","product_rating", "overall_rating", "brand", 
           "product_specifications"
          ]

delete_redundant_columns(columns, amazon)
delete_redundant_columns(columns, flipkart)

In [9]:
amazon.head()

product_name  retail_price  discounted_price
0    Alisha Solid Women's Cycling Shorts           982               438
1    FabHomeDecor Fabric Double Sofa Bed         32143             29121
2                             AW Bellies           991               551
3    Alisha Solid Women's Cycling Shorts           694               325
4  Sicons All Purpose Arnica Dog Shampoo           208               258

In [10]:
flipkart.head()

product_name  retail_price  discounted_price
0    Alisha Solid Women's Cycling Shorts         999.0             379.0
1    FabHomeDecor Fabric Double Sofa Bed       32157.0           22646.0
2                             AW Bellies         999.0             499.0
3    Alisha Solid Women's Cycling Shorts         699.0             267.0
4  Sicons All Purpose Arnica Dog Shampoo         220.0             210.0

In [11]:
def get_lowercase_text(column: str, data: pd.DataFrame) -> None:
    
    """ converts capital/uppercase strings or text into lowercase. """
    
    if column is None:
        print(f"empty column name")
        return 
    if data.shape[0] == 0:
        print(f"empty dataset: {len(data)}")
    
    data[column] = data[column].apply(lambda x: " ".join(x.lower() for x in x.split()))
    
    return

In [12]:
get_lowercase_text('product_name', flipkart)
get_lowercase_text('product_name', amazon)

In [13]:
amazon.head()

product_name  retail_price  discounted_price
0    alisha solid women's cycling shorts           982               438
1    fabhomedecor fabric double sofa bed         32143             29121
2                             aw bellies           991               551
3    alisha solid women's cycling shorts           694               325
4  sicons all purpose arnica dog shampoo           208               258

In [14]:
flipkart.head()

product_name  retail_price  discounted_price
0    alisha solid women's cycling shorts         999.0             379.0
1    fabhomedecor fabric double sofa bed       32157.0           22646.0
2                             aw bellies         999.0             499.0
3    alisha solid women's cycling shorts         699.0             267.0
4  sicons all purpose arnica dog shampoo         220.0             210.0

In [15]:
# flipkart dataset where we can see null example in total --
# There are total 78 missing/null values in each columns retail_price and discounted_price.

# flipkart[flipkart.isna().any(axis=1)]

In [16]:
amazon.rename(columns={"product_name": "amz_product", 
                       "retail_price": "amz_retail_price", 
                       "discounted_price": "amz_discounted_price"
                      }, 
              inplace = True
             )

## Hypothesis:

- As mentioned in the problem statement that these two datasets has their products in common.
- What I want to do here is to make sure that; there are any similar products in given datasets.



- So my hypothesis here is that (also mentioned in the problem statement) there are similar products in given two datasets.
- If not we reject our hypothesis.


## Proof:

- To make sure that the given datasets has similar products, I'll use set data datastructure to check my hypothesis, I will save both dataset's product names saparately and use intersection method on those two sets to get all the common products names.
- Reason behind using set is because rather than iterating through all the products and comparing one by one can lead to great computation and time complexity, instead set methods are efficient and great alternative in case of our problem.


- Since there's a lot of repetition of same product, we are going to get only one occurance of that product from both the datasets, as set intersection method gives us only the common elements from two collections.

In [17]:
amazon_prod_set = set(amazon.amz_product.to_list())
flipkart_prod_set = set(flipkart.product_name.to_list())

# -- Gives all the common products from both sets -- 
# list(amazon_prod_set.intersection(flipkart_prod_set))   

len(amazon_prod_set.intersection(flipkart_prod_set))

12604

### Note:

- Here we can see that there are total 12604 unique products that are common in both datasets, and these products are repeating themselves again and again yet with different retatil and discounted prices.
- We've proved aur hypothesis.


- Below I'm quickly going to check some products which can be available in above two sets.

In [18]:
def check_product(product_name: str, products_set: set) -> bool:
    
    """ returns a boolean value if product is present or not """
    
    if product_name == "":
        print("enter valid product name.")
        return
    if len(products_set) == 0:
        print("empty set given.")
        return
    
    if product_name in products_set:
        return True
    
    return False

In [19]:
product_names = ["alisha solid women's cycling shorts", 
                 "alisha solid women's cycling ", 
                 "faith leenise 1000137 women heels"
                ]

for prod in product_names:
    print(f"{prod} : amazon products set : {check_product(prod, amazon_prod_set)}\n"
          f"{prod} : flipkart products set : {check_product(prod, amazon_prod_set)}\n"
         )

alisha solid women's cycling shorts : amazon products set : True
alisha solid women's cycling shorts : flipkart products set : True

alisha solid women's cycling  : amazon products set : False
alisha solid women's cycling  : flipkart products set : False

faith leenise 1000137 women heels : amazon products set : True
faith leenise 1000137 women heels : flipkart products set : True



## Combining amazon & flipkart dataset for comparison:

In [20]:
combined_data = amazon.copy()

combined_data['flipk_product'] = flipkart['product_name']
combined_data['flipk_retail_price'] = flipkart['retail_price']
combined_data['flip_discounted_price'] = flipkart['discounted_price']

# added this one extra column to see the difference between the retail prices of two datasets.

combined_data['retail_diff'] = np.where(combined_data['amz_retail_price'] == flipkart['retail_price'], 
                                        0,
                                        np.abs(combined_data['amz_retail_price'] - flipkart['retail_price'])
                                       )

combined_data.head()

amz_product  amz_retail_price  \
0    alisha solid women's cycling shorts               982   
1    fabhomedecor fabric double sofa bed             32143   
2                             aw bellies               991   
3    alisha solid women's cycling shorts               694   
4  sicons all purpose arnica dog shampoo               208   

   amz_discounted_price                          flipk_product  \
0                   438    alisha solid women's cycling shorts   
1                 29121    fabhomedecor fabric double sofa bed   
2                   551                             aw bellies   
3                   325    alisha solid women's cycling shorts   
4                   258  sicons all purpose arnica dog shampoo   

   flipk_retail_price  flip_discounted_price  retail_diff  
0               999.0                  379.0         17.0  
1             32157.0                22646.0         14.0  
2               999.0                  499.0          8.0  
3               699.0                  267.0          5.0  
4               220.0                  210.0         12.0

In [21]:
# Null data that is came along with filpkart data.

# combined_data[combined_data.isna().any(axis=1)]

## Results:

In [22]:
combined_data[combined_data['amz_product'].str.contains("muchmore alloy copper charm bracelet")]

amz_product  amz_retail_price  \
16762  muchmore alloy copper charm bracelet                -1   
16763  muchmore alloy copper charm bracelet               592   
16764  muchmore alloy copper charm bracelet               686   
16935  muchmore alloy copper charm bracelet               796   

       amz_discounted_price                         flipk_product  \
16762                     0  muchmore alloy copper charm bracelet   
16763                   349  muchmore alloy copper charm bracelet   
16764                   379  muchmore alloy copper charm bracelet   
16935                   425  muchmore alloy copper charm bracelet   

       flipk_retail_price  flip_discounted_price  retail_diff  
16762                 NaN                    NaN          NaN  
16763               600.0                  300.0          8.0  
16764               699.0                  300.0         13.0  
16935               798.0                  330.0          2.0

In [23]:
combined_data[combined_data['flipk_product'].str.contains("muchmore alloy copper charm bracelet")]

amz_product  amz_retail_price  \
16762  muchmore alloy copper charm bracelet                -1   
16763  muchmore alloy copper charm bracelet               592   
16764  muchmore alloy copper charm bracelet               686   
16935  muchmore alloy copper charm bracelet               796   

       amz_discounted_price                         flipk_product  \
16762                     0  muchmore alloy copper charm bracelet   
16763                   349  muchmore alloy copper charm bracelet   
16764                   379  muchmore alloy copper charm bracelet   
16935                   425  muchmore alloy copper charm bracelet   

       flipk_retail_price  flip_discounted_price  retail_diff  
16762                 NaN                    NaN          NaN  
16763               600.0                  300.0          8.0  
16764               699.0                  300.0         13.0  
16935               798.0                  330.0          2.0

In [24]:
combined_data[combined_data['amz_product'].str.contains("fdt women's leggings")]

amz_product  amz_retail_price  amz_discounted_price  \
28  fdt women's leggings pants               698                   362   

           flipk_product  flipk_retail_price  flip_discounted_price  \
28  fdt women's leggings               699.0                  309.0   

    retail_diff  
28          1.0

## Notes:

- As we can see that there is a repetition of a product multiple times but the retail price and discounted price is seen different at each occurances.
- There are __null values__ as well with these products because we have seen total __78 null__ values in flipkart (retail price and discounted price columns), which we have combined here to compare the products and there prices.

- I cannot impute the missing the values in repective fields, since there were no instructions according to this subject.


- Two things I consider, When we have missing data or values in dataset.  
    1. Data Engineering team should work on that if available.
    2. As a data scientist I will talk to the seniors, or data eng. team, understand reasons behind the missing data or the missing data shows a pattern that we could take into account and then I'd proceed further with my work.



__So here I'm leaving missing data as it is__

In [25]:
combined_data.shape

(20000, 7)

In [26]:
def fetch_product_info(product_name: str) -> pd.DataFrame:
    global combined_data
    
    product_name = product_name.lower()
    
    return combined_data[combined_data['amz_product'].str.contains(product_name)]

In [27]:
fetch_product_info("women's")

amz_product  amz_retail_price  \
0                    alisha solid women's cycling shorts               982   
3                    alisha solid women's cycling shorts               694   
6                    alisha solid women's cycling shorts              1198   
9                    alisha solid women's cycling shorts              1197   
11                                carrel printed women's              2281   
...                                                  ...               ...   
19786                  my addiction women's a-line dress              1293   
19787                       indibox women's a-line dress              1483   
19788                         jiiah women's sheath dress              1043   
19812  lord's red women's slingback peeptoes women heels              1280   
19881  lord's antique gold women's peeptoe heels wome...              1634   

       amz_discounted_price  \
0                       438   
3                       325   
6                       602   
9                       542   
11                     1080   
...                     ...   
19786                   934   
19787                  1136   
19788                   904   
19812                   784   
19881                  2001   

                                           flipk_product  flipk_retail_price  \
0                    alisha solid women's cycling shorts               999.0   
3                    alisha solid women's cycling shorts               699.0   
6                    alisha solid women's cycling shorts              1199.0   
9                    alisha solid women's cycling shorts              1199.0   
11                                carrel printed women's              2299.0   
...                                                  ...                 ...   
19786                  my addiction women's a-line dress              1299.0   
19787                       indibox women's a-line dress              1499.0   
19788                         jiiah women's sheath dress              1049.0   
19812  lord's red women's slingback peeptoes women heels              1299.0   
19881  lord's antique gold women's peeptoe heels wome...              1650.0   

       flip_discounted_price  retail_diff  
0                      379.0         17.0  
3                      267.0          5.0  
6                      479.0          1.0  
9                      479.0          2.0  
11                     910.0         18.0  
...                      ...          ...  
19786                  779.0          6.0  
19787                  999.0         16.0  
19788                  734.0          6.0  
19812                  650.0         19.0  
19881                 1650.0         16.0  

[3789 rows x 7 columns]